### ISSUE TITLE
HA and the NLB

### ISSUE THREAD

**Issue by pkdavidson** [Esri/arcgis-cookbook#35]
We're on a pure VM system with lots of virtual "Windows Server 2012 R2" servers running around.
It's my understanding that we can use IIS 7.5 to basically build up a virtual NLB.  For example:
[https://technet.microsoft.com/en-us/library/cc740265(v=ws.10).aspx](url)
[http://virtuallyhyper.com/2013/04/load-balancing-iis-sites-with-nlb/](url)

Basically looks like one can use the NLB now built into Microsoft's server software.
Will this type of setup work with your proposed HA setup?  
Or do the tomcat, postgresql, etc... components prohibit IIS from controlling the balancing?
Does that even make sense?  

It looks to me like we can we start by deploying just the primary ha site and then as we have time, introduce the secondary site.  Looking over the Json, it looks like the primary site is basically the same as the standard webgis-windows.json file with the exception of adding in a domain name and of course a file server for sharing the config stores for ArcGIS server and Portal.
(That's the piece that has been giving new VM webGIS farms fits over the past year, isn't it?)

I also notice that the webgis-windows-ha-primary.json file does not have the final step to cook up federation.  Is that an oversight or something that requires doing that by hand?

Many diagrams show the NLB as a separate VM "gateway" into the NLB'd system
[http://dotnetcodr.com/2013/06/17/web-farms-in-net-and-iis-part-1-a-general-introduction/](url)

and I've also heard the web adaptor referred to as an NLB in some contexts.
Would it be best to setup a separate VM and put 4 web adaptors on it, 2 for HA-Primary and 2 for HA-Secondary?  Then install IIS there and use it as your NLB?
One could also put your fileserver on this box or is this going to overloading the VM (just toss more resources at it?)

Talking with our system guys, I guess one issue you face is that you have to keep Primary and Secondary boxes tuned identically.  i.e. same map/feature services, layers, etc.... 

We weren't sure NLBs were even available for a Virtual setting (other than the obvious IIS software ones)
However, I did find some virtual NLBs:
[https://kemptechnologies.com/server-load-balancing-appliances/virtual-load-balancers/vlm-overview/](url)
[http://loadbalancer.org//products/virtual](url)

Curious as to your thoughts about all of this.
Not exactly a trivial thing to get going but looks quite powerful if it all comes together.

**Comment by pbobov** [comment]
You are approaching an uncharted territory...

You can configure Web Adaptor to load balance between two or more server or portal machines, but this will not be an HA solution, because the Web Adaptor machine will be a single point of failure.

For true HA deployments you will need either hardware or software NLB. Software NLBs, like NLB clusters that comes with Windows Server 2102, make the network hub to broadcast traffic to all machines in the cluster, and the machines chose who processes requests. For best performance two NIC cards are required. A few years ago I tried to configure NLB cluster from Windows Server 2012 VMs, but had to give up, because this required the machines to have static IP addresses and a broadcast IP. I could not do it on my VMs. My experience with software NLBs ends there. 

The virtual NLB approach looks interesting. This might be a good option for virtualized environments. I have never tried them, though... 

Federation is the last recipe in the webgis-windows-ha-secondary.json. You can move it to webgis-windows-ha-primary.json if you need to postpone deployment of secondary machine.

**Comment by pbobov** [comment]
Here is how Web GIS HA deployed with webgis-windows-ha-*.json looks like in the context of Amazon AWS. The data is synced between machines automatically. If primary machine fails, the secondary becomes primary.

![Web GIS HA](https://s3.amazonaws.com/arcgisstore104/5524/templates/G71815_Optional-additonal-datastore_Diagram.jpg)

**Comment by pkdavidson** [comment]
Dang, I hate it when I miss the obvious like the Federation step in the secondary json.
I also now have the theme to Jaws stuck in my head.  :-)

Seems like picking up a Virtual NLB for use would be the way to go.  
The ones from LoadBalancer look interesting
Could try IIS/Win NLB but it looks messy but maybe that's the way to try initially (free.)
The beauty of Chef and the VM world is that I can always toss out Machine 2 and start again.

Thanks for the info!  I think I'll install the primary and then go from there.
But i'll have to look over the recipes for the Federation step in secondary json file.

How is the data synched automatically between Machine 1 & 2?
Scripts? or is that a benefit of AWS or your ELB?
ELB? Enterprise Load Balancer?  Electronic?  ;-)

I'll have some questions on the Spatiotemporal Big Data Store...
e.g. Is it registered with each AGS on Machine 1 & 2?
I assume it has to be?

Does it make sense to talk about an HA for Machine 3?
This needs to be very solid file server, right?
I wonder if we can use a net app filter that attaches to both Machine 1 and Machine 2.
Have to talk with the System gurus on that.

**Comment by pbobov** [comment]
> How is the data synched automatically between Machine 1 & 2?

Portal and Data Store use PostrgreSQL configured with master-slave replication. Portal content and Server config store are stored on the file server. 

>Is it registered with each AGS on Machine 1 & 2?

Yes. The data store is registered with ArcGIS Server. It can be registered with any ArcGIS Server machine (because the machines share configuration).

> Does it make sense to talk about an HA for Machine 3?

Yes. You do need a highly available file server for HA.

**Comment by pkdavidson** [comment]
Do the Chef HA recipes/json setups do the PostreSQL master-slave settings for us when we install the secondary box?  
Or is it a manual tweak after the installs?

Thanks.

**Comment by pbobov** [comment]
The master-slave replication on PostgreSQL is configured not by Chef, but by data store and portal themselves. Chef calls some admin APIs on portal (e.g. create portal site on primary, and join portal site on secondary). PostgreSQL is embedded and is completely managed by data store and portal.

**Comment by pkdavidson** [comment]
Ah yes, that is one of the new features of 10.4, built in HA?

**Comment by pbobov** [comment]
Actually, it was in 10.3.1 as well. Just in 10.3.1 it was a lot more complex to configure.

**Comment by pkdavidson** [comment]
About to fire off our Production Portal.
Going back to your prior statemet:
_"Federation is the last recipe in the webgis-windows-ha-secondary.json. You can move it to webgis-windows-ha-primary.json if you need to postpone deployment of secondary machine. "_

I shouldn't have any problem with first federating my Production Portal in the webgis-windows-ha-primary.json file and then later creating the secondary ha box and redoing the federation?

And on another question:  
If at a later date we were to move our backup_dir, directories_root and content_dir from one network share out to another location, if I were to then just adjust the json file and rerun Chef, would that take care of what would be needed to move the shared folders that are so crucial to the HA setup?

I am installing to a shared file server but we are undergoing NetApp upgrades and so some naming changes might be occurring down the road.

Finally, are you giving a talk at the UC on the Chef Cookbook?
I think you should.  I'm always surprised that not every admin out there has jumped all over this.

I think setting up my production Portal is going to take me 30 min or so, maybe a bit more.
But that would be an impressive demo to give.  Talk about Chef and have the cookbook ready to go, start at the beginning of your talk, show a new portal at the end...  Everyone in the room has spent a week or two getting it to just be stood up would be amazed.  At least they should :-)

**Comment by pkdavidson** [comment]
Just pulled an error regarding installing NetFx3.
I'm assuming this must be related to the HA install because my prior Dev Portal install was windows only and it ran without a hitch?  Or else I did the 3.5 framework install by hand and have just forgotten.

I'm assuming I just have to do the install for 3.5 by hand via the Server Manager?  
(After I dig up my iso for Srvr2012R2)

**Comment by pbobov** [comment]
If you run 'federation' recipe on another machine with the same attributes as on the first one, the recipe should do nothing. The server by that time will be already federated. The only problem I can think of is that after federation server authentication/authorization setting will be changed to federated. Theoretically, it should not be a problem because chef only uses server's super admin account that does not change. 

The cookbook currently does not support changing directories except for Data Store backup directory. So far, you will have to do it manually.

We already presented the first version of ArcGIS cookbook at UC 2015. At UC 2016 we also have a session about ArcGIS Automation.

Good luck with your production deployment!

**Comment by pbobov** [comment]
On Windows 8.1 NetFx3 must be installed manually (it is not included in OS). But we did not have this problem on 2012 R2.

**Comment by pkdavidson** [comment]
Wonder how I missed the UC2015 demo.
Guess I wasn't Chef aware yet.

That's odd regarding Fx3.5 because I am installing to 2012 R2
But I don't recall having to do it by hand with the Dev install.
Oh well, I just did it now and will see if that clears up the issue.
Looks like it did, I'm off to installing AGS 10.4 so that's past where it hung up.

We did clone a box I had put ArcDesktop onto and then from the clone made a VM template.
Maybe something went south in that....  
So far so good, thanks for all your help.

**Comment by blingenfelder** [comment]
Did you end up trying the Network Load Balance through IIS?  I am currently setting up a Dev environment trying to do the same as it seems that Esri is no longer recommending using multiple clusters in a single site.  Clustering our environment impacted performance quite a bit even when adding resources to the machines, which is why I'm looking into this option.  Just wondering if anyone else has had success with this setup.

**Comment by pkdavidson** [comment]
Sorry Bling... I have not tried NLB via IIS (yet.)  And yes, the word I've heard is a single site is better, at least with VMs?   

Pavel:  not sure what happened but my file server share went away at the same time my Prod Portal disk overflowed....  Sort of a comedy of errors and a mistake on my part.  We had a 500GB disk on Prod but I'd forgotten to extend the partition from 50 to 500 so....

I think the thing to do is to move the config files to the local drive rather than trying to share.
I found when the share goes away, everything screeches to a halt.  Not surprising....

Since we really had nothing going on there yet and since my backup was not created, I'm thinking it might be easiest to recreate the site?  Might require a new licensing file from Esri but that's quick enough.

If I can recover the file share then I'll drop it back in place and see about moving the config folders around by hand. But I'm not finding info on that process.  I find info on moving the content folder but not the config folders.

After going through this exercise, I'm thinking that an HA setup should have the config folders on the local machine and the SA box would then point to those folders (which will have been setup as a share.)

Thoughts on this?  I suppose an HA setup should have an HA file server (thought I did...)

**Comment by pbobov** [comment]
Ouch. Consider this as a test of disaster recovery plan. 

If you don't yet have any valuable data in the store, then recreating the deployment is probably the easiest way out, but eventually you will have data you wouldn't want to loose. 

Using primary machine as a file server is as good as using a standalone machine. It might even improve performance a bit.

You probably should look at backing up your data (see
http://server.arcgis.com/en/portal/latest/administer/windows/overview-backup-restore-web-gis.htm, http://server.arcgis.com/en/server/latest/administer/windows/back-up-and-restore-your-arcgis-server-site-configuration.htm).

**Comment by pkdavidson** [comment]
Turns out the NIC on the share had locked up...  We're in the process of migrating to new NetApp appliances so might be related...

I had backed up our Test server but not Prod since it was pretty empty.
However, I had done some unique configurations so I should have been backing up.

Interestingly, the Data Store is not happy and the service will restart and then immediately shut down.  talked with an Esri Engineer in Tx who thought the best thing was to uninstall DS, delete folders and then recreate it.  I'll probably do that to see if I can bring it backup.  But then I'll blow it all away and either restore a snapshot of the VM or rebuild it all.  
I also learned that there are issues with using straigh Windows shares for the shared folders as well as with NAS.  So I'll start with a local folder and if/when we build up HA, I'll look to do a CIFS/SAN setup.

Like you say, good DR practice.  (not the rebuilding...)  And now I need to script up a backup scenario.
All in all, a very painless lesson given what it could have been.
